From 9485326cdc97ef197f597159de8b0a4534008f97 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stephan=20M=C3=BCller?= <smueller@chronox.de>
Date: Sun, 21 Nov 2021 15:31:27 +0100
Subject: [PATCH 0077/1646] [Backport] crypto: rsa - limit key size to 2048 in
 FIPS mode

mainline inclusion
from mainline-v5.17-rc1
commit 1ce1bacc480965fab4420e561916ce45d2e90c05
category: feature
bugzilla: 188668

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=1ce1bacc480965fab4420e561916ce45d2e90c05

--------------------------------

FIPS disallows RSA with keys < 2048 bits. Thus, the kernel should
consider the enforcement of this limit.

Signed-off-by: Stephan Mueller <smueller@chronox.de>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Signed-off-by: Cui GaoSheng <cuigaosheng1@huawei.com>
---
 crypto/rsa.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/crypto/rsa.c b/crypto/rsa.c
index ece7bafd6984..144808d61776 100644
--- a/crypto/rsa.c
+++ b/crypto/rsa.c
@@ -5,6 +5,7 @@
  * Authors: Tadeusz Struk <tadeusz.struk@intel.com>
  */
 
+#include <linux/fips.h>
 #include <linux/module.h>
 #include <linux/mpi.h>
 #include <crypto/internal/rsa.h>
@@ -144,6 +145,9 @@ static int rsa_check_key_length(unsigned int len)
 	case 512:
 	case 1024:
 	case 1536:
+		if (fips_enabled)
+			return -EINVAL;
+		fallthrough;
 	case 2048:
 	case 3072:
 	case 4096:
-- 
2.17.1

